core/src/main/java/org/apache/calcite/rex/RexUtil.java

```diff
@@ -2750,4 +2750,27 @@ public ExprSimplifier(RexSimplify simplify, RexUnknownAs unknownAs,
       return simplify.rexBuilder.makeCast(call.getType(), simplifiedNode, matchNullability);
     }
   }
+
+  /**
+   * Utility to fix the inconsistencies of nullable/non-nullable types.
+   */
+  public static class RexInputRefNullabilityFixer extends RexShuttle {
+
+    private final RexBuilder rexBuilder;
+
+    private final RelDataType desiredTypes;
+
+    public RexInputRefNullabilityFixer(RexBuilder rexBuilder, RelDataType desiredTypes) {
+      this.rexBuilder = rexBuilder;
+      this.desiredTypes = desiredTypes;
+    }
```

Comment:
This class is a little confusing. Could you explain more about why we should have it?

Reply:
Sure. 

For a RexNode, the nullability type of the underlying RexInputRef may be incorrect. 
For example, given a RexCall ADD($1, 2), the type of the input reference $1 may be nullable, while the underlying data type should be non-nullable.

This inconsistency may cause some assert to fail, making the program break.

It seems the fundamental solution to this problem is to generate the correct nullability type from the very beginning. However, it is not always possible to do so, because for some scenarios, the nullability type of the input reference is not known when the RexNode is being created. 

So we introduce this utility to fix the nullability type. It may seem a little hack. However, it is feasible, and provides a clean solution.

Reply:
`FixNullabilityShuttle` seems to meet your requirements.

Reply:
Revised. Thank you for the good suggestion.